### PR TITLE
Expectation ftp v20

### PR DESCRIPTION
**Pull request by regit**
Update of #3002 

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/550

Describe changes:
- Fix SCLogDebug call

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/352
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/136

### CODE REVIEW

File: src/app-layer-expectation.c [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+#define EXPECTATION_TIMEOUT 30
+
+typedef struct Expectation_ {
+    struct timeval ts;
+    Port sp;
+    Port dp;
+    AppProto alproto;
+    int direction;
+    void *data;
+    void (*DFree)(void *);
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: is this to free the data field? Comment!

File: src/app-layer-expectation.c [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    /* As we are creating the expectation, we release lock on IPPair without
+     * setting the ref count to 0. This way the IPPair will be kept till
+     * cleanup */
+    IPPairUnlock(ipp);
+    return 0;
+
+error:
+    if (exp != NULL)
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: exp can't be NULL AFAICS

File: src/app-layer-expectation.c [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+     */
+    (void) IPPairDecrUsecnt(ipp);
+    SC_ATOMIC_SUB(expectation_count, 1);
+    if (pexp == NULL) {
+        IPPairSetStorageById(ipp, g_expectation_id, lexp);
+    } else {
+        pexp->next = lexp;
+    }
+    if (exp->data)
+        SCFree(exp->data);
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: Is this the field that should be freed by the DFree func ptr?

File: src/app-layer-expectation.c [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+    int x = SC_ATOMIC_GET(expectation_count);
+    if (x == 0) {
+        return ALPROTO_UNKNOWN;
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
+    time_t ctime = f->lastts.tv_sec;
+
+    if (exp == NULL)
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: seems to make sense to check this right after the 'Get'?

File: src/app-layer-expectation.c [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    IPPair *ipp = NULL;
+    Expectation *lexp = NULL;
+    Expectation *pexp = NULL;
+
+    int x = SC_ATOMIC_GET(expectation_count);
+    if (x == 0) {
+        return ALPROTO_UNKNOWN;
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: add comment like: "'takes reference of the ip pair in 'ipp'"

File: src/app-layer-expectation.c [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    pexp = NULL;
+    while (exp) {
+        lexp = exp->next;
+        if ( (exp->direction & direction) &&
+             ((exp->sp == 0) || (exp->sp == f->sp)) &&
+             ((exp->dp == 0) || (exp->dp == f->dp))) {
+            alproto = exp->alproto;
+            f->alproto_ts = alproto;
+            f->alproto_tc = alproto;
+            if (FlowSetStorageById(f, g_expectation_data_id, exp->data) != 0) {
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: comment like: transfer ownership of exp->data to flow

File: src/app-layer-expectation.c [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
+    time_t ctime = f->lastts.tv_sec;
+
+    if (exp == NULL)
+        goto out;
+
+    pexp = NULL;
+    while (exp) {
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: it seems that if we have multiple expectations that apply to a flow each of them would be used to update the flow. Would FlowSetStorageById handle that case w/o loosing memory?

File: src/app-layer-ftp.c [hunk 8]
```diff
@@ -228,31 +272,121 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
...
+                state->port_line_len = state->current_line_len;
+                break;
+            case FTP_COMMAND_RETR:
+                /* change direction (default to server) so expectation will handle
+                 * the correct message when expectation will match.
+                 */ 
+                direction = STREAM_TOCLIENT;
+                // fallthrough
+            case FTP_COMMAND_STOR:
+                {
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: can you add a comment that if we get here 'current_line_len' is at least >=5? This is the type of code that looks suspicious as it doesn't check things itself.

I would also not be opposed to actually checking the side here, purely as a defensive measure so future changes to FTPGetLine/FTPParseRequestCommand don't cause unintended issues

File: src/app-layer-ftp.c [hunk 9]
```diff
@@ -228,31 +272,121 @@ static int FTPParseRequest(Flow *f, void *ftp_state,
...
+                    struct FtpTransferCmd *data = SCCalloc(1, sizeof(struct FtpTransferCmd));
+                    if (data == NULL)
+                        SCReturnInt(-1);
+                    data->DFree = FtpTransferCmdFree;
+                    data->filename = SCCalloc(state->current_line_len - 4, sizeof(char));
+                    if (data->filename == NULL) {
+                        SCFree(data);
+                        SCReturnInt(-1);
+                    }
+                    memcpy(data->filename, state->current_line + 5, state->current_line_len - 5);
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: it seems the goal here is to create a nul-terminated string. Two things:

1. please make this explicit
2. is this guaranteed to be a valid string? As it comes from the wire my suspicion is 'no'. So a byte array instead?

File: src/app-layer-ftp.c [hunk 10]
```diff
@@ -446,6 +795,32 @@ void RegisterFTPParsers(void)
...
+        AppLayerParserRegisterGetTx(IPPROTO_TCP, ALPROTO_FTPDATA, FTPDataGetTx);
+
+        AppLayerParserRegisterGetTxCnt(IPPROTO_TCP, ALPROTO_FTPDATA, FTPDataGetTxCnt);
+
+        AppLayerParserRegisterGetStateProgressFunc(IPPROTO_TCP, ALPROTO_FTPDATA, FTPDataGetAlstateProgress);
+
+        AppLayerParserRegisterGetStateProgressCompletionStatus(ALPROTO_FTPDATA,
+                FTPDataGetAlstateProgressCompletionStatus);
+
+        sbcfg.buf_size = 4096;
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: seems the file data itself is excluded from the memcap. To include it register alloc/free funcs here for the streaming api

File: src/app-layer-ftp.c [hunk 11]
```diff
@@ -465,6 +840,34 @@ void FTPAtExitPrintStats(void)
 #endif
 }
 
+
+#ifdef HAVE_LIBJANSSON
+json_t *JsonFTPDataAddMetadata(const Flow *f)
+{
+    FtpDataState *ftp_state = NULL;
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: const

File: src/app-layer-ftp.h [hunk 12]
```diff
@@ -131,15 +132,39 @@ typedef struct FtpState_ {
...
     DetectEngineState *de_state;
 } FtpState;
 
+enum {
+    FTPDATA_STATE_IN_PROGRESS,
+    FTPDATA_STATE_FINISHED,
+};
+
+/** FTP Data State for app layer parser */
+typedef struct FtpDataState_ {
```
Comment (victorjulien) [inline src/app-layer-ftp.h]: I think the layout can be optimized

File: src/app-layer-expectation.c [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+    } else if (FLOW_IS_IPV6(f)) {
+        FLOW_COPY_IPV6_ADDR_TO_PACKET(&f->src, ip_src);
+        FLOW_COPY_IPV6_ADDR_TO_PACKET(&f->dst, ip_dst);
+    } else {
+        return -1;
+    }
+    return 0;
+}
+
+static Expectation *AppLayerExpectationGet(Flow *f, int direction, IPPair **ipp)
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: AppLayerExpectationLookup? It wraps around IPPairLookupIPPairFromHash, not IPPairGetIPPairFromHash

File: src/app-layer-expectation.c [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+ *
+ * \return expectation data identifier
+ */
+int AppLayerExpectationGetDataId(void)
+{
+    return g_expectation_data_id;
+}
+
+static Expectation * RemoveExpectationAndGetNext(IPPair *ipp,
+                                Expectation *pexp, Expectation *exp,
```
Comment (victorjulien) [inline src/app-layer-expectation.c]: what are the names pexp and lexp referring to?

File: src/app-layer-ftp.c (hunk 8 above)
Comment (victorjulien) [reply src/app-layer-ftp.c]: *checking the size

File: src/app-layer-expectation.c (hunk 1 above)
Comment (regit) [reply src/app-layer-expectation.c]: Just updated that part of the code to only get the DFree in ExpectationData.

File: src/app-layer-expectation.c (hunk 14 above)
Comment (regit) [reply src/app-layer-expectation.c]: Adding documentation to the function.

File: src/app-layer-ftp.c (hunk 9 above)
Comment (regit) [reply src/app-layer-ftp.c]: From https://tools.ietf.org/html/rfc959 at section 5.3.2 we are suppose to have ASCII only here. I'm adding a check here. Mister @chifflier do you know if I should be more laxist here ?

Comment (victorjulien) [reply src/app-layer-ftp.c]: Making it a byte array makes it match behavior of our other protocol parsers and gives consistent matching logic in our keywords. So this is what I think we should do.

Comment (regit) [reply src/app-layer-ftp.c]: OK, will do that.

File: src/app-layer-ftp.c (hunk 10 above)
Comment (regit) [reply src/app-layer-ftp.c]: I got the idea but don't manage to understand what to do here. Can you tell me which functions to use ?

Comment (victorjulien) [reply src/app-layer-ftp.c]: see https://github.com/OISF/suricata/blob/master/src/app-layer-htp.c#L2274

The idea is to register your own alloc/free funcs that track memuse/memcap

Comment (regit) [reply src/app-layer-ftp.c]: Do you want me to introduce a new memcap or should I use the reassembly one ?

Comment (victorjulien) [reply src/app-layer-ftp.c]: Not sure what reassembly has to do with it.

Whats needed is a FTP specific memcap.

Comment (regit) [reply src/app-layer-ftp.c]: Ok :+1:

**Comment by regit** [comment]
Gonna resubmit.

**Pull request closed without merging**